README.md
Pyroute2
========

Pyroute2 is a pure Python **netlink** library. The core requires only Python
stdlib, no 3rd party libraries. The library was started as an RTNL protocol
implementation, so the name is **pyroute2**, but now it supports many netlink
protocols. Some supported netlink families and protocols:

* **rtnl**, network settings --- addresses, routes, traffic controls
* **nfnetlink** --- netfilter API:

    * **ipset** --- IP sets
    * **nftables** --- packet filtering
    * **nfct** --- connection tracking

* **ipq** --- simplest userspace packet filtering, iptables QUEUE target
* **devlink** --- manage and monitor devlink-enabled hardware
* **generic** --- generic netlink families:

    * **ethtool** --- low-level network interface setup
    * **wireguard** --- VPN setup
    * **nl80211** --- wireless functions API (basic support)
    * **taskstats** --- extended process statistics
    * **acpi_events** --- ACPI events monitoring
    * **thermal_events** --- thermal events monitoring
    * **VFS_DQUOT** --- disk quota events monitoring

On the low level the library provides socket objects with an
extended API. The additional functionality aims to:

* Help to open/bind netlink sockets
* Discover generic netlink protocols and multicast groups
* Construct, encode and decode netlink and PF_ROUTE messages

Supported systems
-----------------

Pyroute2 runs natively on Linux and emulates some limited subset
of RTNL netlink API on BSD systems on top of PF_ROUTE notifications
and standard system tools.

Other platforms are not supported.

NDB -- high level RTNL API
--------------------------

Key features:

* Data integrity
* Transactions with commit/rollback changes
* State synchronization
* Multiple sources, including netns and remote systems

A "Hello world" example::

    from pyroute2 import NDB

    ndb = NDB(log='on')
    for record in ndb.interfaces.summary():
        print(record)

    for record in ndb.addresses.summary():
        print(record)

    (ndb
     .interfaces
     .create(ifname='br0', kind='bridge')  # create a bridge
     .add_port('eth0')                     # add ports
     .add_port('eth1')                     #
     .add_ip('10.0.0.1/24')                # add addresses
     .add_ip('192.168.0.1/24')             #
     .set('br_stp_state', 1)               # set STP on
     .set('state', 'up')                   # bring the interface up
     .commit())                            # commit pending changes

IPRoute -- Low level RTNL API
-----------------------------

Low-level **IPRoute** utility --- Linux network configuration.
The **IPRoute** class is a 1-to-1 RTNL mapping. There are no implicit
interface lookups and so on.

Get notifications about network settings changes with IPRoute::

    from pyroute2 import IPRoute
    with IPRoute() as ipr:
        # With IPRoute objects you have to call bind() manually
        ipr.bind()
        for message in ipr.get():
            print(message)

More examples::

    from socket import AF_INET
    from pyroute2 import IPRoute

    # get access to the netlink socket
    ip = IPRoute()
    # no monitoring here -- thus no bind()

    # print interfaces
    for link in ip.get_links():
        print(link)

    # create VETH pair and move v0p1 to netns 'test'
    ip.link('add', ifname='v0p0', peer='v0p1', kind='veth')
    idx = ip.link_lookup(ifname='v0p1')[0]
    ip.link('set', index=idx, net_ns_fd='test')

    # bring v0p0 up and add an address
    idx = ip.link_lookup(ifname='v0p0')[0]
    ip.link('set', index=idx, state='up')
    ip.addr('add', index=idx, address='10.0.0.1', prefixlen=24)

    # release Netlink socket
    ip.close()

Network namespace examples
--------------------------

Network namespace manipulation::

    from pyroute2 import netns
    # create netns
    netns.create('test')
    # list
    print(netns.listnetns())
    # remove netns
    netns.remove('test')

Create **veth** interfaces pair and move to **netns**::

    from pyroute2 import IPRoute

    with IPRoute() as ipr:

        # create interface pair
        ipr.link('add', ifname='v0p0', kind='veth',  peer='v0p1')

        # lookup the peer index
        idx = ipr.link_lookup(ifname='v0p1')[0]

        # move the peer to the 'test' netns:
        ipr.link('set', index='v0p1', net_ns_fd='test')

List interfaces in some **netns**::

    from pyroute2 import NetNS
    from pprint import pprint

    ns = NetNS('test')
    pprint(ns.get_links())
    ns.close()

More details and samples see in the documentation.

Installation
------------

`make install` or `pip install pyroute2`

Requirements
------------

Python >= 2.7

The pyroute2 testing and documentaion framework requirements:

* flake8
* coverage
* nosetests
* sphinx
* aafigure
* netaddr
* dtcd (optional, https://github.com/svinota/dtcd)

Optional dependencies:

* mitogen -- for distributed rtnl
* psutil -- for ss2 tool

Links
-----

* home: https://pyroute2.org/
* srcs: https://github.com/svinota/pyroute2
* bugs: https://github.com/svinota/pyroute2/issues
* pypi: https://pypi.python.org/pypi/pyroute2
* docs: http://docs.pyroute2.org/
* list: https://groups.google.com/d/forum/pyroute2-dev


CHANGELOG.md
Changelog
=========

* 0.6.13
    * requests: IPv6 routes default priority <https://github.com/svinota/pyroute2/issues/934>
    * ndb: fix NetlinkError escalation routine on apply <https://github.com/svinota/pyroute2/issues/934>
* 0.6.12
    * ndb: allow dict as keys in `__contains__`
    * ndb: fixed `count()` on nested views
    * ndb: fix IPv6 addr removal: <https://github.com/svinota/pyroute2/issues/926>
    * ifinfmsg: fixed IFLA_IPTUN_FLAGS: <https://github.com/svinota/pyroute2/pull/924>
    * tc: fix tcm_info setting: <https://github.com/svinota/pyroute2/issues/928>
* 0.6.11
    * ipdb: fix requests compatibility: <https://github.com/svinota/pyroute2/issues/920>
    * ipdb: added deprecation warning
* 0.6.10
    * ndb: fix DB reload on incomplete requests in the IPRoute API fallbacks: <https://github.com/svinota/pyroute2/issues/890>
    * ndb: basic ipvlan support
    * ndb: support `ndb.schema.backup("filename")` fro SQLite3 backend
    * ndb: transactions, CheckProcess, PingAddress, see `test_ndb/test_transaction.py` and docs
    * ndb: make snapshots only for `commit()` calls, no automatic snapshots for `apply()`
    * netlink: filter out all the subsequent messages starting from the first error: <https://github.com/svinota/pyroute2/issues/892>
    * protocols: fix arp module export
    * iproute: deprecate positional arguments to `IPRoute.rule()`
    * iproute, ndb: use `pyroute2.requests` to filter and transform API call arguments
    * conntrack: fix exports: <https://github.com/svinota/pyroute2/pull/900>
    * nftables: multiple fixes: <https://github.com/svinota/pyroute2/pull/902>
    * tc: fix em_ipset crash: <https://github.com/svinota/pyroute2/pull/905>
    * tests: integrated pre-commit, github actions and more
    * tests: support basic OpenBSD tests
    * pyroute2-cli: parser fixes
* 0.6.9
    * wireguard peer fix for both IPv4 and IPv6: <https://github.com/svinota/pyroute2/issues/888>
    * netns: avoid extra libc reloads: <https://github.com/svinota/pyroute2/pull/887>
* 0.6.8
    * wireguard: IPv6 fixes: <https://github.com/svinota/pyroute2/pull/882>
    * ndb: support arbitrary IPRoute-compatible source classes
    * ndb: tolerate source failures
    * ndb: chaotic tests
    * general: kernel version parser: <https://github.com/svinota/pyroute2/issues/885>
    * package: static version file: <https://github.com/svinota/pyroute2/discussions/884>
    * iproute: support custom link kind classes: <https://github.com/svinota/pyroute2/issues/886>
* 0.6.7
    * ndb: reload of tunnels in the down state <https://github.com/svinota/pyroute2/issues/878>
    * ndb: improved reliability in the state sync, affected multihop and encap routes
    * ndb: improved neighbours management
* 0.6.6
    * tuntap: support loongarch <https://github.com/svinota/pyroute2/pull/824>
    * ndb: fix routes cleanup <https://github.com/svinota/pyroute2/issues/849>
    * ndb: support `in` for sources, interfaces and addresses
    * ndb: support changing tunnel (gre, ipip, ...) attributes
    * netlink: support NLM_F_DUMP_INTR <https://github.com/svinota/pyroute2/issues/874>
* 0.6.5
    * ndb: fix `in` for sources: <https://github.com/svinota/pyroute2/issues/843>
    * core: provide entry_points aliases for legacy code: <https://github.com/svinota/pyroute2/issues/845>
    * packaging: fix missing exports: <https://github.com/svinota/pyroute2/issues/816>
* 0.6.4
    * ndb: memory leak fix: <https://github.com/svinota/pyroute2/issues/789>
    * packaging: fix internal modules import
    * netlink: DEFAULT_RCVBUF size increased: <https://github.com/svinota/pyroute2/issues/813>
    * rtnl: fix COLLECT_METADATA for geneve devices: <https://github.com/svinota/pyroute2/issues/795>
* 0.6.3
    * ndb: route spec fixes: <https://github.com/svinota/pyroute2/issues/808>
    * packaging: force virtual packages to require same versions
* 0.6.2
    * core: relocate RemoteIPRoute: <https://github.com/svinota/pyroute2/issues/798>
    * wireguard: fix key parsing: <https://github.com/svinota/pyroute2/pull/804>
    * packaging: fix importlib-metadata deps: <https://github.com/svinota/pyroute2/issues/797>
    * tc: support pfifo qdisc: <https://github.com/svinota/pyroute2/issues/801>
* 0.6.1
    * packaging: new layout: <https://github.com/svinota/pyroute2/discussions/786>
* 0.5.19
    * ndb: fix default routes keys: <https://github.com/svinota/pyroute2/issues/790>
    * mptcp: basic support: <https://github.com/svinota/pyroute2/issues/782>
    * netlink: ext_ack support fix: <https://github.com/svinota/pyroute2/pull/759>
    * tc: netem rate setting: <https://github.com/svinota/pyroute2/pull/780>
    * NSPopen: fix fd leak: <https://github.com/svinota/pyroute2/issues/622>
* 0.5.18
    * netlink: support zero length lladdr: <https://github.com/svinota/pyroute2/issues/717>
* 0.5.17
    * license: aligned cli/ss2 <https://github.com/svinota/pyroute2/pull/777>
    * ndb: `del_ip()` improvements
    * ndb: `wait(timeout=...)`
* 0.5.16
    * ndb: fix syntax for Python < 3.6 <https://github.com/svinota/pyroute2/pull/776>
* 0.5.15
    * ndb: don't mess with SQL adapters unless really needed
    * ndb: support more virtual interface types
    * ndb: support `in` for views and record sets
    * ndb: fix nested fetches from DB
    * tests: start migration to pytest
* 0.5.14
    * iproute: ip neigh get <https://github.com/svinota/pyroute2/pull/723>
    * iproute: link_lookup fix <https://github.com/svinota/pyroute2/pull/725>
    * nftables: missing hooks <https://github.com/svinota/pyroute2/pull/732>
    * netns: fix <https://github.com/svinota/pyroute2/pull/733>
* 0.5.13
    * netns: allow to pass custom libc reference <https://github.com/svinota/pyroute2/issues/702>
    * generic: att L2TP support <https://github.com/svinota/pyroute2/pull/709>
    * iproute: link_lookup() optimization <https://github.com/svinota/pyroute2/issues/712>
    * ndb: basic cluster support
* 0.5.12

pyroute2/arp.py
from pyroute2.common import map_namespace

## ARP protocol HARDWARE identifiers.
ARPHRD_NETROM = 0    # from KA9Q: NET/ROM pseudo
ARPHRD_ETHER = 1    # Ethernet 10Mbps
ARPHRD_EETHER = 2    # Experimental Ethernet
ARPHRD_AX25 = 3    # AX.25 Level 2
ARPHRD_PRONET = 4    # PROnet token ring
ARPHRD_CHAOS = 5    # Chaosnet
ARPHRD_IEEE802 = 6    # IEEE 802.2 Ethernet/TR/TB
ARPHRD_ARCNET = 7    # ARCnet
ARPHRD_APPLETLK = 8    # APPLEtalk
ARPHRD_DLCI = 15    # Frame Relay DLCI
ARPHRD_ATM = 19    # ATM
ARPHRD_METRICOM = 23    # Metricom STRIP (new IANA id)
ARPHRD_IEEE1394 = 24    # IEEE 1394 IPv4 - RFC 2734
ARPHRD_EUI64 = 27    # EUI-64
ARPHRD_INFINIBAND = 32    # InfiniBand

## Dummy types for non ARP hardware
ARPHRD_SLIP = 256
ARPHRD_CSLIP = 257
ARPHRD_SLIP6 = 258
ARPHRD_CSLIP6 = 259
ARPHRD_RSRVD = 260    # Notional KISS type
ARPHRD_ADAPT = 264
ARPHRD_ROSE = 270
ARPHRD_X25 = 271    # CCITT X.25
ARPHRD_HWX25 = 272    # Boards with X.25 in firmware
ARPHRD_PPP = 512
ARPHRD_CISCO = 513    # Cisco HDLC
ARPHRD_HDLC = ARPHRD_CISCO
ARPHRD_LAPB = 516    # LAPB
ARPHRD_DDCMP = 517    # Digital's DDCMP protocol
ARPHRD_RAWHDLC = 518    # Raw HDLC

ARPHRD_TUNNEL = 768    # IPIP tunnel
ARPHRD_TUNNEL6 = 769    # IP6IP6 tunnel
ARPHRD_FRAD = 770    # Frame Relay Access Device
ARPHRD_SKIP = 771    # SKIP vif
ARPHRD_LOOPBACK = 772    # Loopback device
ARPHRD_LOCALTLK = 773    # Localtalk device
ARPHRD_FDDI = 774    # Fiber Distributed Data Interface
ARPHRD_BIF = 775    # AP1000 BIF
ARPHRD_SIT = 776    # sit0 device - IPv6-in-IPv4
ARPHRD_IPDDP = 777    # IP over DDP tunneller
ARPHRD_IPGRE = 778    # GRE over IP
ARPHRD_PIMREG = 779    # PIMSM register interface
ARPHRD_HIPPI = 780    # High Performance Parallel Interface
ARPHRD_ASH = 781    # Nexus 64Mbps Ash
ARPHRD_ECONET = 782    # Acorn Econet
ARPHRD_IRDA = 783    # Linux-IrDA
## ARP works differently on different FC media .. so
ARPHRD_FCPP = 784    # Point to point fibrechannel
ARPHRD_FCAL = 785    # Fibrechannel arbitrated loop
ARPHRD_FCPL = 786    # Fibrechannel public loop
ARPHRD_FCFABRIC = 787    # Fibrechannel fabric
## 787->799 reserved for fibrechannel media types
ARPHRD_IEEE802_TR = 800    # Magic type ident for TR
ARPHRD_IEEE80211 = 801    # IEEE 802.11
ARPHRD_IEEE80211_PRISM = 802    # IEEE 802.11 + Prism2 header
ARPHRD_IEEE80211_RADIOTAP = 803    # IEEE 802.11 + radiotap header
ARPHRD_MPLS_TUNNEL = 899    # MPLS Tunnel Interface

ARPHRD_VOID = 0xFFFF    # Void type, nothing is known
ARPHRD_NONE = 0xFFFE    # zero header length

(ARPHRD_NAMES, ARPHRD_VALUES) = map_namespace("ARPHRD_", globals())
